transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,641 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{provider::Provider, transport::SocketFamily, ConnectError, Connection, Error};
+
+use bytes::BytesMut;
+use futures::{
+    channel::{mpsc, oneshot},
+    prelude::*,
+};
+use quinn_proto::VarInt;
+use std::{
+    collections::HashMap,
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    ops::ControlFlow,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+use x509_parser::nom::AsBytes;
+
+// The `Driver` drops packets if the channel to the connection
+// or transport is full.
+// Set capacity 10 to avoid unnecessary packet drops if the receiver
+// is only very briefly busy, but not buffer a large amount of packets
+// if it is blocked longer.
+const CHANNEL_CAPACITY: usize = 10;
+
+/// Config for the transport.
+#[derive(Clone)]
+pub struct Config {
+    /// Timeout for the initial handshake when establishing a connection.
+    /// The actual timeout is the minimum of this an the [`Config::max_idle_timeout`].
+    pub handshake_timeout: Duration,
+    /// Maximum duration of inactivity in ms to accept before timing out the connection.
+    pub max_idle_timeout: u32,
+    /// Period of inactivity before sending a keep-alive packet.
+    /// Must be set lower than the idle_timeout of both
+    /// peers to be effective.
+    ///
+    /// See [`quinn_proto::TransportConfig::keep_alive_interval`] for more
+    /// info.
+    pub keep_alive_interval: Duration,
+    /// Maximum number of incoming bidirectional streams that may be open
+    /// concurrently by the remote peer.
+    pub max_concurrent_stream_limit: u32,
+
+    /// Max unacknowledged data in bytes that may be send on a single stream.
+    pub max_stream_data: u32,
+
+    /// Max unacknowledged data in bytes that may be send in total on all streams
+    /// of a connection.
+    pub max_connection_data: u32,
+
+    /// TLS client config for the inner [`quinn_proto::ClientConfig`].
+    client_tls_config: Arc<rustls::ClientConfig>,
+    /// TLS server config for the inner [`quinn_proto::ServerConfig`].
+    server_tls_config: Arc<rustls::ServerConfig>,
+}
+
+impl Config {
+    /// Creates a new configuration object with default values.
+    pub fn new(keypair: &libp2p_core::identity::Keypair) -> Self {
+        let client_tls_config = Arc::new(libp2p_tls::make_client_config(keypair, None).unwrap());
```

Comment:
Originally we added the `None` parameter here to be able to specify the expected `PeerId` and have the handshake fail instead of aborting the connection afterwards on mismatch.

The config is shared here though so we have to set it to `None`. This would require some refactoring to be able to use this feature.